src/algebra/direct_sum_graded.lean

```diff
@@ -543,6 +545,98 @@ end comm_ring
 
 end grade_zero
 
+section to_semiring
+
+variables {R : Type*} [Π i, add_comm_monoid (A i)] [add_monoid ι] [gmonoid A] [semiring R]
+variables {A}
+
+/-- If two ring homomorphisms from `⨁ i, A i` are equal on each `of A i y`,
+then they are equal.
+
+See note [partially-applied ext lemmas]. -/
+@[ext]
+lemma ring_hom_ext' (F G : (⨁ i, A i) →+* R)
+  (h : ∀ i, (F : (⨁ i, A i) →+ R).comp (of _ i) = (G : (⨁ i, A i) →+ R).comp (of _ i)) : F = G :=
+ring_hom.coe_add_monoid_hom_injective $ direct_sum.add_hom_ext' h
+
+/-- A family of `add_monoid_hom`s preserving `direct_sum.ghas_one.one` and `direct_sum.ghas_mul.mul`
+describes a `ring_hom`s on `⨁ i, A i`. This is a stronger version of `direct_sum.to_monoid`.
+
+Of particular interest is the case when `A i` are bundled subojects, `f` is the family of
+coercions such as `add_submonoid.subtype (A i)`, and the `[gmonoid A]` structure originates from
+`direct_sum.gmonoid.of_add_submonoids`, in which case the proofs about `ghas_one` and `ghas_mul`
+can be discharged by `rfl`. -/
+@[simps]
+def to_semiring
+  (f : Π i, A i →+ R) (hone : f _ (ghas_one.one) = 1)
+  (hmul : ∀ {i j} (ai : A i) (aj : A j), f _ (ghas_mul.mul ai aj) = f _ ai * f _ aj) :
+  (⨁ i, A i) →+* R :=
+{ to_fun := to_add_monoid f,
+  map_one' := begin
+    change (to_add_monoid f) (of _ 0 _) = 1,
+    simp only [to_add_monoid_of],
+    exact hone
+  end,
+  map_mul' := λ x y, begin
+    change (to_add_monoid f).comp (add_monoid_hom.mul_left x) y =
+      (add_monoid_hom.mul_left $ (to_add_monoid f) x).comp (to_add_monoid f) y,
+    apply add_monoid_hom.congr_fun,
+    ext yi yv : 2,
+    let y' := direct_sum.of _ yi yv,
+    change (to_add_monoid f).comp (add_monoid_hom.mul_right y') x =
+      (add_monoid_hom.mul_right $ (to_add_monoid f) y').comp (to_add_monoid f) x,
+    apply add_monoid_hom.congr_fun,
+    ext xi xv : 2,
+    let x' := direct_sum.of _ xi xv,
```

Comment:
This could be a little shorter if we had a `add_monoid_hom.mul : R →+ R →+ R`, but we don't and our import order makes it diffficult to put that anywhere near `add_monoid_hom.mul_left` and `add_monoid_hom.mul_right`.

Reply:
Added in a bad place in #7382